File: camerax_android/lib/src/ml/common/input_image_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class InputImageImpl implements InputImage {
  final InputImageProxyApi api;

  InputImageImpl.internal(this.api);

  @override
  Future<int> getWidth() => api.getWidth();
  @override
  Future<int> getHeight() => api.getHeight();
}

final class InputImageChannelImpl extends InputImageChannel {
  @override
  InputImage fromFilePath(Uri imageUri) {
    final api = InputImageProxyApi.fromFilePath(imageUri: '$imageUri');
    return InputImageImpl.internal(api);
  }
}

extension InputImageX on InputImage {
  InputImageProxyApi get api {
    final impl = this;
    if (impl is! InputImageImpl) throw TypeError();
    return impl.api;
  }

  Future<Size<int>> getSize() async {
    final width = await getWidth();
    final height = await getHeight();
    return Size(width, height);
  }
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/ml/common/InputImageImpl.kt
package dev.zeekr.camerax_android.ml.common

import com.google.mlkit.vision.common.InputImage
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PigeonApiInputImageProxyApi
import dev.zeekr.camerax_android.common.uriImpl
import dev.zeekr.camerax_android.context

class InputImageImpl(private val registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiInputImageProxyApi(registrar) {
    override fun fromFilePath(imageUri: String): InputImage {
        return InputImage.fromFilePath(registrar.context, imageUri.uriImpl)
    }

    override fun getWidth(pigeon_instance: InputImage): Long {
        return pigeon_instance.width.toLong()
    }

    override fun getHeight(pigeon_instance: InputImage): Long {
        return pigeon_instance.height.toLong()
    }
}